src/level_set/FESurfaceDistanceEvaluator.cpp

```diff
@@ -0,0 +1,1310 @@
+                        }
+#endif
+#if (NDIM == 3)
+                        proj = getClosestPoint3D(P, n0, n1, n2);
+                        dist = (P - proj).norm();
+#endif
+
+                        // If the distance is the same as the minimal distance, then the cell is equidistant
+                        // to multiple elements, so add it to the set.
+                        if (MathUtilities<double>::equalEps(dist, min_dist))
+                        {
+                            vec_equidistant_pair.push_back(std::make_pair(elem, proj));
+                        }
+                        else if (dist < min_dist)
+                        {
+                            // If a new minimal element is found, clear the previous vector of pairs and simply keep
+                            // this one
+                            min_dist = dist;
+                            vec_equidistant_pair.clear();
+                            vec_equidistant_pair.push_back(std::make_pair(elem, proj));
+                        }
+                    }
+                }
+
+                // Determine the sign based on the normal vectors at the quadrature points
+                // For linear elements considered in this class, the normal is the same no matter the quadrature point
+                // If the cell is equidistant to multiple elements, take the average normal from those elements
+                double sgn = 0.0;
+                const size_t vec_length = vec_equidistant_pair.size();
+                TBOX_ASSERT(vec_length > 0);
+                IBTK::VectorNd avg_unit_normal, avg_proj;
+                avg_unit_normal.setZero();
+                avg_proj.setZero();
+                std::vector<std::pair<Elem*, IBTK::VectorNd> >::const_iterator it;
+                for (it = vec_equidistant_pair.begin(); it != vec_equidistant_pair.end(); ++it)
+                {
+                    Elem* elem = (*it).first;
+                    IBTK::VectorNd proj = (*it).second;
+                    avg_proj += proj;
+
+                    // Loop over the sides of the element. If it has no neighbors on a side,
+                    // then it MUST live on the boundary of the mesh
+                    for (unsigned int s = 0; s < elem->n_sides(); ++s)
+                    {
+                        if (elem->neighbor(s) != NULL) continue;
+                        {
+                            fe_bdry->reinit(elem, s);
+                            const unsigned int n_qp = qrule_bdry->n_points();
+                            const std::vector<libMesh::Point>& bdry_normals = fe_bdry->get_normals();
+                            TBOX_ASSERT(n_qp == 1); // for now
+                            unsigned int qp = 0;
+                            for (int d = 0; d < NDIM; ++d)
+                            {
+                                avg_unit_normal(d) += bdry_normals[qp](d);
+                            }
```

Comment:
Since this class is of interest, let's use the normal in the right way. We should follow the angle weighted normal as explained here:
https://ieeexplore.ieee.org/document/1407857
https://www.sciencedirect.com/science/article/pii/S0021999114004495

@drwells : Would like to implement this?